### PR TITLE
cleanup: prevent panic in cleanUpSnapshot

### LINKED ISSUE #2192: panic when creating thick pvc-pvc clone
```
I0618 11:47:58.720223       1 rbd_util.go:997] ID: 23 Req-ID: pvc-4e1baf07-0a5a-444f-a1c0-7e59884ecd5a setting disableInUseChecks: false image features: [layering] mounter: rbd
I0618 11:47:58.728419       1 omap.go:84] ID: 23 Req-ID: pvc-4e1baf07-0a5a-444f-a1c0-7e59884ecd5a got omap values: (pool="replicapool", namespace="", name="csi.volume.e6ef75ad-d029-11eb-8023-0242ac110006"): map[csi.imageid:308281eb56dd csi.imagename:csi-vol-e6ef75ad-d029-11eb-8023-0242ac110006 csi.volname:pvc-46fa615c-453a-4fa9-8fbf-0d41bfe47118 csi.volume.owner:default]
I0618 11:47:58.788534       1 omap.go:84] ID: 23 Req-ID: pvc-4e1baf07-0a5a-444f-a1c0-7e59884ecd5a got omap values: (pool="replicapool", namespace="", name="csi.volumes.default"): map[csi.volume.pvc-4e1baf07-0a5a-444f-a1c0-7e59884ecd5a:4bb9f33b-d02a-11eb-8023-0242ac110006]
I0618 11:47:58.793371       1 omap.go:84] ID: 23 Req-ID: pvc-4e1baf07-0a5a-444f-a1c0-7e59884ecd5a got omap values: (pool="replicapool", namespace="", name="csi.volume.4bb9f33b-d02a-11eb-8023-0242ac110006"): map[csi.imageid:308279f47f9b csi.imagename:csi-vol-4bb9f33b-d02a-11eb-8023-0242ac110006 csi.volname:pvc-4e1baf07-0a5a-444f-a1c0-7e59884ecd5a csi.volume.owner:default]
I0618 11:47:58.840550       1 rbd_journal.go:330] ID: 23 Req-ID: pvc-4e1baf07-0a5a-444f-a1c0-7e59884ecd5a found existing volume (0001-0009-rook-ceph-0000000000000001-4bb9f33b-d02a-11eb-8023-0242ac110006) with image name (csi-vol-4bb9f33b-d02a-11eb-8023-0242ac110006) for request (pvc-4e1baf07-0a5a-444f-a1c0-7e59884ecd5a)
E0618 11:47:58.885251       1 utils.go:182] panic occurred: runtime error: invalid memory address or nil pointer dereference
goroutine 89 [running]:
runtime/debug.Stack(0x0, 0x0, 0x4186b8)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x25
github.com/ceph/ceph-csi/internal/csi-common.panicHandler.func1(0xc00056d8d8)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:183 +0x10a
panic(0x1a0c920, 0x2c2a280)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/ceph/ceph-csi/internal/rbd.(*rbdVolume).deleteSnapshot(0xc0006b2c40, 0x1f1a720, 0xc0006cdf50, 0x0, 0x0, 0x0)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/internal/rbd/rbd_util.go:1077 +0x79
github.com/ceph/ceph-csi/internal/rbd.cleanUpSnapshot(0x1f1a720, 0xc0006cdf50, 0xc0006b2c40, 0x0, 0xc0006b2a80, 0xc0006bd680, 0x0, 0x0)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/internal/rbd/snapshot.go:65 +0x67
github.com/ceph/ceph-csi/internal/rbd.(*ControllerServer).repairExistingVolume(0xc00000f7e0, 0x1f1a720, 0xc0006cdf50, 0xc0004ace70, 0xc0006bd680, 0xc0006b2a80, 0xc0006b2c40, 0x0, 0x2, 0xc0006dd000, ...)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/internal/rbd/controllerserver.go:389 +0x27b
github.com/ceph/ceph-csi/internal/rbd.(*ControllerServer).CreateVolume(0xc00000f7e0, 0x1f1a720, 0xc0006cdf50, 0xc0004ace70, 0x0, 0x0, 0x0)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/internal/rbd/controllerserver.go:285 +0xda5
github.com/container-storage-interface/spec/lib/go/csi._Controller_CreateVolume_Handler.func1(0x1f1a720, 0xc0006cdf50, 0x1b7d1c0, 0xc0004ace70, 0x16f44d9, 0x2cbe0c0, 0xc000000000, 0x0)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:5507 +0x89
github.com/ceph/ceph-csi/internal/csi-common.panicHandler(0x1f1a720, 0xc0006cdf50, 0x1b7d1c0, 0xc0004ace70, 0xc0006bd320, 0xc0006bd340, 0x0, 0x0, 0x0, 0x0)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:187 +0x96
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1f1a720, 0xc0006cdf50, 0x1b7d1c0, 0xc0004ace70, 0xc00057d988, 0x1, 0x1, 0x417d50)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x63
github.com/ceph/ceph-csi/internal/csi-common.logGRPC(0x1f1a720, 0xc0006cdf50, 0x1b7d1c0, 0xc0004ace70, 0xc0006bd320, 0xc0006bd360, 0x1b219e0, 0xc0006cdf50, 0x197f4e0, 0xc0006d74c0)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:170 +0x1f1
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1f1a720, 0xc0006cdf50, 0x1b7d1c0, 0xc0004ace70, 0x1985860, 0xc0006d74d0, 0x1f1a720, 0xc0006cdf50)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x63
github.com/ceph/ceph-csi/internal/csi-common.contextIDInjector(0x1f1a720, 0xc0006cdf20, 0x1b7d1c0, 0xc0004ace70, 0xc0006bd320, 0xc0006bd380, 0x170831a, 0x1adb580, 0xc0006bd3a0, 0xc0006bd320)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:159 +0x122
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1f1a720, 0xc0006cddd0, 0x1b7d1c0, 0xc0004ace70, 0xc000512800, 0x0, 0xc00057db30, 0x4186b8)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x1f1a720, 0xc0006cddd0, 0x1b7d1c0, 0xc0004ace70, 0xc0006bd320, 0xc0006bd340, 0xc000515ba0, 0x51a706, 0x1b219e0, 0xc0006cddd0)
	/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34 +0xd7
github.com/container-storage-interface/spec/lib/go/csi._Controller_CreateVolume_Handler(0x1bc0480, 0xc00000f7e0, 0x1f1a720, 0xc0006cddd0, 0xc0006cbc80, 0xc0003c5c80, 0x1f1a720, 0xc0006cddd0, 0xc0005f6c00, 0x1f8)
```

Pulled the master code and build image and this got caused when doing the below steps

* Create 10 Gib Thick PVC
* Create 2 parallel clones of 10Gib each
* restart csi-provisioner pod when cloning is going on

### PR DESCRIPTION
**Pull request by nixpanic**
While cleaning up snapshots, not all object may exist after a partial
provisioning attempt. In case objects are missing, do not try to delete
them.

Fixes: #2192

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e-helm/k8s-1.19

Failed to start the minikube VM:
```
* Deleting "minikube" in kvm2 ...

! StartHost failed, but will try again: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

* Creating kvm2 VM (CPUs=8, Memory=14336MB, Disk=20000MB) ...

* Failed to start kvm2 VM. Running "minikube delete" may fix it: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute


X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
```

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e-helm/k8s-1.19

Failed to deploy ceph-csi ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.19/detail/mini-e2e-helm_k8s-1.19/1620/pipeline/)):
```
Error from server: etcdserver: request timed out
```

**Review by yati1998 — COMMENTED** [review]
LGTM

**Review by Madhu-1 — DISMISSED** [review]

File: internal/rbd/snapshot.go [hunk 1]
```diff
@@ -61,21 +61,29 @@ func createRBDClone(ctx context.Context, parentVol, cloneRbdVol *rbdVolume, snap
...
-	err := parentVol.deleteSnapshot(ctx, rbdSnap)
-	if err != nil {
-		if !errors.Is(err, ErrSnapNotFound) {
-			util.ErrorLog(ctx, "failed to delete snapshot: %v", err)
-			return err
+	if parentVol != nil && rbdSnap != nil {
+		err := parentVol.deleteSnapshot(ctx, rbdSnap)
+		if err != nil {
+			if !errors.Is(err, ErrSnapNotFound) {
+				util.ErrorLog(ctx, "failed to delete snapshot: %v", err)
```
Comment (Madhu-1) [inline internal/rbd/snapshot.go]: maybe log the snapshot name here?

File: internal/rbd/snapshot.go [hunk 2]
```diff
@@ -61,21 +61,29 @@ func createRBDClone(ctx context.Context, parentVol, cloneRbdVol *rbdVolume, snap
...
-	if err != nil {
-		if !errors.Is(err, ErrImageNotFound) {
-			util.ErrorLog(ctx, "failed to delete rbd image: %s/%s with error: %v", rbdVol.Pool, rbdVol.VolName, err)
-			return err
+
+	if rbdVol != nil {
+		err := deleteImage(ctx, rbdVol, cr)
+		if err != nil {
+			if !errors.Is(err, ErrImageNotFound) {
+				util.ErrorLog(ctx, "failed to delete rbd image: %s/%s with error: %v", rbdVol.Pool, rbdVol.VolName, err)
```
Comment (Madhu-1) [inline internal/rbd/snapshot.go]: use `rbdVol` for logging as the String interface is implemented.

**Review by Madhu-1 — COMMENTED** [review]
Suggestion to improve as this PR is not changing this part this can be fixed in a separate PR also.

File: internal/rbd/snapshot.go (hunk 1 above)
Comment (nixpanic) [reply internal/rbd/snapshot.go]: done

File: internal/rbd/snapshot.go (hunk 2 above)
Comment (nixpanic) [reply internal/rbd/snapshot.go]: done

**Review by Madhu-1 — APPROVED** [review]

**Review by yati1998 — APPROVED** [review]

**Review by Rakshith-R — APPROVED** [review]

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Review by Yuggupta27 — APPROVED** [review]

**Review by humblec — APPROVED** [review]

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.20

Failed due to #1969 ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.20/detail/mini-e2e_k8s-1.20/1298/pipeline))

**Pull request merged**